### ISSUE TITLE
Layerlist-bundle hierarchy development

### ISSUE THREAD

**Issue by mr-lev1** [oskariorg/oskari-docs#234]
The new React-based layerlist-bundle is great. The only things missing are the hierarchy features as seen in hierarchical-layerlist -bundle. We are starting a project that aims in adding all hierarchical-layerlist features to the new, React-based layerlist-bundle. 

In a nutshell this feature allows the users to create sub-groups for groups (also called themes) in the layer list and to define which layers belong to the groups/sub-groups. The maximum depth of the hierarchy is 3, so a group can have sub-groups and a sub-group can have sub-groups.

Also, we will be adding the possibility to toggle groups (show/hide layers) and re-order groups/layers by dragging.

### Development project goals:
- Implement sub-group adding (a plus icon next to group settings button)
- Implement a layer tree in layer settings (for selecting groups/sub-groups to which a layer belong to)
- Implement toggle for layer groups (checkbox, show/hide all layer in a group)
- Implement warning when toggling a group with X or more layers (this value will be configurable in properties, default 10)
- Implement re-ordering of groups/layers by dragging

**Comment by ZakarFin** [comment]
Good stuff! But I do have a couple of concerns:

1) We have groups with 400+ layers on them so the checkbox  to add/remove all layers in group to/from map should be optional. Maybe add a toggle to the group if the checkbox should be shown or not?

2) On the technical side I'm very hesitant on how to implement the dragging layers/groups for setting their order in the hierarchy properly in this case. The admin-functionalities are currently very nicely contained on the admin/admin-layereditor bundle. The current layerlist bundle just provides enough hooks for the admin-functionalities to integrate on the layer listing UI. I would _very much_ want to keep it that way. The ordering by dragging will quite easily bleed too much admin-functionality to the layer listing. This would not only make the application more bloated for non-admin users, but makes the whole listing functionality more complex than it needs to be (makes it harder to maintain in the future and might easily make the application more "laggy" for end-users). We considered the hierarchical setup for the React implementation a bit when we were designing it and one thought was to have some UI component in the group insert/add popup where the admin can choose the parent for the group. But that doesn't solve how to set the order of layers/groups really. But I'm excited to see how you would solve these 👍 Make sure to communicate/PR often on this so we can see where it's headed earlier rather than later.

3) I'm not sure if there's a need to technically restrict the structure to be just 3 levels deep. In the current hierarchical-layerlist implementation the loops are hard-coded but you could just use recursion instead. The UI might break at some point for too deep structure but I'd still make the code function without hard-coded inner loops.

Those are the main concerns that I have for this change. Other than that this is a good improvement to have!

**Comment by mr-lev1** [comment]
Thank you very much for your comments, @ZakarFin!

Here are some comments for your concerns:

1. Indeed. It seems I did not mention this in my original issue posting, but we will be implementing the mentioned toggle for the group selection. This feature was also included in the original hierarchical-layerlist (in the group settings dialog). We will implement it in a similar way.

2. We discussed this with the team in the beginning and also recognized that it might not be that simple to implement just by modifying admin-layereditor -bundle. We have not yet figured out how to do this, but will keep you posted :) We will be in touch about this!

3. This is also a good point and sounds reasonable to not limit the depth.
